### ISSUE TITLE
[Phase 3] AI Model Integration - GPU-Accelerated Vision Service with PyTorch

### ISSUE THREAD

**Issue by zordhalo** [zordhalo/photo-organizer#3]
## 🎯 Objective
Implement GPU-accelerated vision service using PyTorch for construction photo classification.

### 1. Vision Service Setup
- [ ] Create `backend/services/vision_service.py`
- [ ] Implement GPU detection and device allocation
- [ ] Add fallback to CPU if GPU unavailable
- [ ] Configure CUDA memory management

### 2. Model Integration
- [ ] Load pre-trained ResNet50 from torchvision
- [ ] Implement model initialization with weights
- [ ] Add model to GPU/CPU device
- [ ] Configure model for inference mode (eval())

### 3. Image Preprocessing Pipeline
- [ ] Create image transformation pipeline (resize, normalize)
- [ ] Implement batch preprocessing for multiple images
- [ ] Add validation for image formats and sizes
- [ ] Handle EXIF orientation correction

### 4. Inference Implementation
- [ ] Single image inference method
- [ ] Batch inference with GPU optimization
- [ ] Post-processing for classification results
- [ ] Confidence score calculation

### 5. Construction Categories Mapping
- [ ] Define construction-specific categories:
  - Foundation & Excavation
  - Framing & Structure
  - Roofing
  - Electrical & Plumbing
  - Interior Finishing
  - Exterior & Landscaping
  - Safety & Equipment
  - Progress Documentation
- [ ] Map ImageNet classes to construction categories
- [ ] Create fallback keyword detection system

## 🔧 Technical Implementation

```python
import torch
import torchvision.models as models
import torchvision.transforms as transforms
from PIL import Image
from typing import List, Dict, Tuple
import logging

class VisionService:
    def __init__(self):
        self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        logging.info(f"Using device: {self.device}")
        
        # Load pre-trained ResNet50
        self.model = models.resnet50(weights=models.ResNet50_Weights.IMAGENET1K_V2)
        self.model.to(self.device)
        self.model.eval()
        
        # Image preprocessing
        self.transform = transforms.Compose([
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize(
                mean=[0.485, 0.456, 0.406],
                std=[0.229, 0.224, 0.225]
            )
        ])
        
    @torch.no_grad()
    def classify_image(self, image: Image.Image) -> Dict[str, any]:
        """Single image classification"""
        input_tensor = self.transform(image).unsqueeze(0).to(self.device)
        output = self.model(input_tensor)
        probabilities = torch.nn.functional.softmax(output[0], dim=0)
        
        top5_prob, top5_idx = torch.topk(probabilities, 5)
        
        return {
            'predictions': [
                {'class': idx.item(), 'confidence': prob.item()}
                for idx, prob in zip(top5_idx, top5_prob)
            ],
            'category': self._map_to_construction_category(top5_idx[0].item())
        }
    
    @torch.no_grad()
    def classify_batch(self, images: List[Image.Image]) -> List[Dict[str, any]]:
        """Batch classification for GPU optimization"""
        batch_tensor = torch.stack([
            self.transform(img) for img in images
        ]).to(self.device)
        
        outputs = self.model(batch_tensor)
        probabilities = torch.nn.functional.softmax(outputs, dim=1)
        
        results = []
        for prob in probabilities:
            top5_prob, top5_idx = torch.topk(prob, 5)
            results.append({
                'predictions': [
                    {'class': idx.item(), 'confidence': p.item()}
                    for idx, p in zip(top5_idx, top5_prob)
                ],
                'category': self._map_to_construction_category(top5_idx[0].item())
            })
        
        return results
    
    def _map_to_construction_category(self, imagenet_class: int) -> str:
        """Map ImageNet classes to construction categories"""
        # Implement mapping logic
        # Return appropriate construction category
        pass
```

## 📊 Performance Targets
- Single image inference: ~50ms on GPU
- Batch inference (10 images): ~11ms per image
- GPU memory usage: <2GB for ResNet50
- Throughput: 90+ images/second in batch mode

## ✅ Acceptance Criteria
- [ ] Vision service successfully loads ResNet50 model
- [ ] GPU acceleration working (when available)
- [ ] Single and batch inference methods implemented
- [ ] Construction category mapping complete
- [ ] Unit tests pass with >80% coverage
- [ ] Performance targets met on GPU hardware

## 📚 References
- [PyTorch Model Deployment](https://www.learnpytorch.io/09_pytorch_model_deployment/)
- [FastAPI with PyTorch](https://nexusml.co/blog/image-classification-inference-fastapi/)
- [GPU Optimization Guide](https://www.labelvisor.com/optimizing-gpu-performance-for-yolov8/)

## ⏱️ Estimated Time
5 hours

## 🔗 Dependencies
- Phase 1: Environment Setup (CUDA, PyTorch installed)
- Phase 2: Core Backend Architecture (FastAPI structure)

## 🏷️ Labels
`backend`, `ai-ml`, `gpu`, `phase-3`
